tempesta_fw/sock_frontend.c

```diff
@@ -33,132 +34,152 @@
+		return r;
 	}
 
-	ss_tcp_set_listen(s, proto);
-	TFW_DBG("Created listening socket %p\n", s->sk);
-
-	/* TODO adjust /proc/sys/net/core/somaxconn */
-	r = s->ops->listen(s, 1024);
+	r = sock_create_kern(addr.sa.sa_family, SOCK_STREAM, IPPROTO_TCP, &s);
 	if (r) {
-		TFW_ERR("Can't listen on front-end socket (%d)\n", r);
-		goto err;
+		TFW_ERR("Can't create socket (err: %d)\n", r);
+		return r;
 	}
 
-	return r;
-err:
-	sock_release(s);
-	return r;
-}
-
-void
-tfw_close_listen_sockets(void)
-{
-	down_read(&tfw_cfg.mtx);
+	inet_sk(s->sk)->freebind = 1;
+	s->sk->sk_reuse = 1;
+	r = s->ops->bind(s, &addr.sa, tfw_addr_sa_len(&addr));
+	if (r) {
+		TFW_ERR("can't bind to address: '%s'\n", addr_str);
+		sock_release(s);
+		return r;
+	}
 
-	TFW_LOG("Close %u listening sockets\n", listen_socks_n);
+	TFW_DBG("Created frontend socket %p\n", s->sk);
 
-	while (listen_socks_n)
-		sock_release(protos[--listen_socks_n].listener);
-	kfree(protos);
+	BUG_ON(listen_socks_n >= ARRAY_SIZE(protos));
+	BUG_ON(listen_socks[listen_socks_n]);
+	listen_socks[listen_socks_n] = s;
+	++listen_socks_n;
 
-	up_read(&tfw_cfg.mtx);
+	return 0;
 }
 
-int
-tfw_open_listen_sockets(void)
+static int
+start_listen_socks(void)
 {
-	struct sockaddr_in6 *addr;
-	int r = -ENOMEM;
-	__be16 ports[DEF_MAX_PORTS];
-
-	down_read(&tfw_cfg.mtx);
-
-	TFW_LOG("Open %u listening sockets\n", tfw_cfg.listen->count);
-
-	protos = kzalloc(sizeof(void *) * tfw_cfg.listen->count, GFP_KERNEL);
-	if (!protos)
-		goto out;
-
-	for (listen_socks_n = 0; listen_socks_n < tfw_cfg.listen->count;
-	     ++listen_socks_n)
-	{
-		SsProto *proto;
-
-		if (listen_socks_n > DEF_MAX_PORTS) {
-			TFW_ERR("Too many listening sockets\n");
-			tfw_close_listen_sockets();
-			goto out;
-		}
+	SsProto *proto;
+	struct socket *sock;
+	int i, r;
 
+	FOR_EACH_SOCK(sock, i) {
 		/*
 		 * TODO If multiprotocol support is required, then here we must
 		 * have information for which protocol we're establishing
 		 * the new listener. So TfwAddrCfg must be extended with
 		 * protocol information (e.g. HTTP enum value).
 		 */
-		proto = protos + listen_socks_n;
+		proto = &protos[i];
 		proto->type = TFW_FSM_HTTP;
-		addr = (struct sockaddr_in6 *)(tfw_cfg.listen->addr
-					       + listen_socks_n);
-		r = __open_listen_socket(proto, addr);
+
+		BUG_ON(proto->listener);
+
+		ss_tcp_set_listen(sock, proto);
+		TFW_DBG("Created listening socket %p\n", sock->sk);
+
+		/* TODO adjust /proc/sys/net/core/somaxconn */
+		r = sock->ops->listen(sock, LISTEN_SOCK_BACKLOG_LEN);
 		if (r) {
-			tfw_close_listen_sockets();
-			goto out;
+			TFW_ERR("Can't listen on front-end socket (%d)\n", r);
+			return r;
 		}
-		ports[listen_socks_n] = addr->sin6_port;
+
 	}
 
-	tfw_filter_set_inports(ports, listen_socks_n);
+	return 0;
+}
 
-	r = 0;
-out:
-	up_read(&tfw_cfg.mtx);
-	return r;
+static void
+stop_listen_socks(void)
+{
+	struct socket *sock;
+	int i;
+
+	FOR_EACH_SOCK(sock, i) {
+		kernel_sock_shutdown(sock, SHUT_RDWR);
```

Comment:
Did you check that the function works with synchronous sockets? IIRC it doesn't work due to special socket callbacks. Moreover, SS has designated ss_close().